### PR TITLE
feat: Add Shadow DOM support for useOverlay and improve event handling

### PR DESCRIPTION
**Pull request by ritz078**
Followup of #6046 

This PR fixes an issue in `useOverlay` where `shouldCloseOnInteractOutside` always received the shadow root parent element as the argument whenever an element inside the shadow root was clicked.

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/WAI/ARIA/apg/)

## 🧢 Your Project:

Nutrient (formerly PSPDFKit)

### CODE REVIEW

**Comment by ritz078** [comment]
@snowystinger Anything I can do to get this one through ?

File: packages/@react-aria/overlays/package.json [hunk 1]
```diff
@@ -28,6 +28,7 @@
     "@react-aria/ssr": "^3.9.7",
     "@react-aria/utils": "^3.27.0",
     "@react-aria/visually-hidden": "^3.8.19",
+    "@react-stately/flags": "^3.0.5",
```
Comment (snowystinger) [inline packages/@react-aria/overlays/package.json:31]: this is just a dev dependency i think

File: packages/@react-aria/overlays/src/useOverlay.ts [hunk 2]
```diff
@@ -101,7 +102,7 @@ export function useOverlay(props: AriaOverlayProps, ref: RefObject<Element | nul
   };
 
   let onInteractOutside = (e: PointerEvent) => {
-    if (!shouldCloseOnInteractOutside || shouldCloseOnInteractOutside(e.target as Element)) {
+    if (!shouldCloseOnInteractOutside || shouldCloseOnInteractOutside(getEventTarget(e))) {
```
Comment (snowystinger) [inline packages/@react-aria/overlays/src/useOverlay.ts:105]: there are some 'relatedTarget' checks below, can we dive into a shadowDOM for those as well? i don't know if there is a composedPath for them

I could see this being an issue if getEventTarget is contained in e.relatedTarget or if the e.relatedTarget is a shadowroot and contains both targets that should be ignored as well as ones which shouldn't.
For example, if someone made a Toast Container inside a shadow dom, then clicking the toasts probably shouldn't count as outside but maybe clicking between individual toasts should for some reason.

File: packages/@react-aria/overlays/test/useOverlay.test.js [hunk 3]
```diff
@@ -29,19 +42,10 @@ function Example(props) {
...
-    ]}
-    ${'Pointer Events'} | ${installPointerEvent}| ${[
-      (el) => fireEvent.pointerDown(el, {button: 0, pointerId: 1}),
-      (el) => fireEvent.pointerUp(el, {button: 0, pointerId: 1})
-    ]}
-    ${'Touch Events'}   | ${() => {}}           | ${[
-      (el) => fireEvent.touchStart(el, {changedTouches: [{identifier: 1}]}),
-      (el) => fireEvent.touchEnd(el, {changedTouches: [{identifier: 1}]})
-    ]}
+    type                | prepare                | actions
```
Comment (snowystinger) [inline packages/@react-aria/overlays/test/useOverlay.test.js:45]: reformatting?

File: packages/@react-aria/overlays/src/useOverlay.ts (hunk 2 above)
Comment (snowystinger) [reply packages/@react-aria/overlays/src/useOverlay.ts:105]: that `isChildOfActiveScope` might be problematic as well if the activescope is within a shadowroot

File: packages/@react-aria/overlays/test/useOverlay.test.js [hunk 4]
```diff
@@ -127,10 +147,90 @@ describe('useOverlay', function () {
...
+        button: 0,
+        pointerId: 1
+      });
       fireEvent.pointerUp(document.body);
       expect(isPrevented).toBeFalsy(); // meaning the event had preventDefault called
     });
   });
 });
+
+describe('useOverlay with shadow dom', () => {
```
Comment (snowystinger) [inline packages/@react-aria/overlays/test/useOverlay.test.js:163]: I don't think these tests are going to be enough, usually overlays portal out to be a direct child of the `body` element, which would actually take them out of a shadowroot potentially

This can also be changed with UNSTABLE_PortalProvider

We'll need tests using the RAC Popover at a minimum

**Review by snowystinger — COMMENTED** [review]
Thanks for your patience, we're in the midst of getting a release out.

File: packages/@react-aria/overlays/src/useOverlay.ts [hunk 5]
```diff
@@ -100,7 +101,7 @@ export function useOverlay(props: AriaOverlayProps, ref: RefObject<Element | nul
...
```
Comment (snowystinger) [inline packages/@react-aria/overlays/src/useOverlay.ts:104]: Is it weird that we call an external API with something that should be inside a shadowroot, it'd unintentionally leak shadow dom internals through our API

We don't do this in the pressEvents, though the use case is obviously a little different

**Review by snowystinger — COMMENTED** [review]
noticed that `isElementInChildOfActiveScope` calls `node.contains` in FocusScope
https://github.com/adobe/react-spectrum/blob/6a35a9a227b5a026a091e9e449bc2d1bb04d3777/packages/%40react-aria/focus/src/FocusScope.tsx#L443
Does this need to use the shadowDOM safe query we added in our utils?

What other tests might we need?
- the non-shadowDOM "firefox bug" describe needs to be checked either in an app or as a test in a shadowDOM environment
- If focus is moving into a child focus scope (e.g. menu inside a dialog)
- Can the underlay be outside the shadowDOM? what other setups might need coverage?
- `shouldCloseOnInteractOutside={(target) => target !== underlay}` what if we don't have access to underlay because the shadowroot is just the component instead of the entire app?

It may be easier to write some of these tests in the React Aria Component file

**Comment by snowystinger** [comment]
This PR's requirements have been met via others
https://github.com/adobe/react-spectrum/pull/9551
https://github.com/adobe/react-spectrum/pull/9485
https://github.com/adobe/react-spectrum/pull/9608
https://github.com/adobe/react-spectrum/pull/9632

Verified by running tests in 9632 where i added the tests written here
